keel-redis/src/main/kotlin/com/netflix/spinnaker/keel/redis/RedisAssetRepository.kt

```diff
@@ -0,0 +1,207 @@
+    withRedis { redis ->
+      redis.smembers(INDEX_SET)
+        .map(::AssetId)
+        .forEach { id ->
+          readAsset(redis, id)
+            ?.also(callback)
+            ?: onInvalidIndex(id)
+        }
+    }
+  }
+
+  override fun get(id: AssetId): Asset? =
+    withRedis { redis ->
+      readAsset(redis, id) as? Asset
+    }
+
+  override fun getPartial(id: AssetId): PartialAsset? =
+    withRedis { redis ->
+      readAsset(redis, id) as? PartialAsset
+    }
+
+  override fun getContainer(id: AssetId): AssetContainer? =
+    withRedis { redis ->
+      val root = readAsset(redis, id) as? Asset
+      if (root != null) {
+        val partials = redis
+          .smembers(id.partialsKey)
+          .asSequence()
+          // TODO: optimally this should be done with a single read operation
+          .map { readAsset(redis, it.let(::AssetId)) as? PartialAsset }
+          .filterNotNull()
+          .toSet()
+        AssetContainer(root, partials)
+      } else {
+        null
+      }
+    }
+
+  override fun store(asset: AssetBase) {
+    withRedis { redis ->
+      redis.hmset(asset.id.key, asset.toHash())
+      if (asset is Asset && asset.dependsOn.isNotEmpty()) {
+        redis.sadd(asset.id.dependsOnKey, *asset.dependsOn.map { it.value }.toTypedArray())
+        asset.dependsOn.forEach {
+          redis.sadd(it.dependenciesKey, asset.id.value)
+        }
+      } else if (asset is PartialAsset) {
+        redis.sadd(
+          asset.root.partialsKey,
+          asset.id.value
+        )
+      }
+      redis.sadd(INDEX_SET, asset.id.value)
+      redis.zadd(asset.id.stateKey, timestamp(), AssetState.Unknown.name)
+    }
+  }
+
+  override fun dependents(id: AssetId): Iterable<AssetId> =
+    withRedis { redis ->
+      redis.smembers(id.dependenciesKey).map(::AssetId)
+    }
+
+  override fun lastKnownState(id: AssetId): Pair<AssetState, Instant>? =
+    withRedis { redis ->
+      redis.zrangeByScoreWithScores(id.stateKey, Double.MIN_VALUE, Double.MAX_VALUE, 0, 1)
+        .asSequence()
+        .map { AssetState.valueOf(it.element) to Instant.ofEpochMilli(it.score.toLong()) }
+        .firstOrNull()
+    }
+
+  override fun updateState(id: AssetId, state: AssetState) {
+    withRedis { redis ->
+      redis.zadd(id.stateKey, timestamp(), state.name)
+    }
+  }
+
+  @PostConstruct
+  fun logKnownAssets() {
+    withRedis { redis ->
+      redis
+        .smembers(INDEX_SET)
+        .sorted()
+        .also { log.info("Managing the following assets:") }
+        .forEach { log.info(" - $it") }
+    }
+  }
+
+  companion object {
+    private const val INDEX_SET = "keel:assets"
+    private const val ASSET_HASH = "keel:asset:%s"
+    private const val DEPENDENCIES_SET = "keel:asset:%s:dependencies"
+    private const val DEPENDS_ON_SET = "keel:asset:%s:dependsOn"
+    private const val PARTIALS_SET = "keel:asset:%s:partials"
+    private const val STATE_SORTED_SET = "keel:asset:%s:state"
```

Comment:
Let's get these hashed' `{keel:asset:%s}` plz, so we can keep dynomite as a viable option.

Reply:
Ultimately I'd just like to do this with SQL but I wanted a quick impl.

Reply:
Yeah, that's fine but it shouldn't prevent us from having a good enough implementation for OSS. The last thing we want is a Front50 Redis situation.

Reply:
Totally agree. This will probably live for a while.